packages/actor-query-operation-group/lib/GroupsState.ts

```diff
@@ -44,56 +49,69 @@ export class GroupsState {
    *
    * @param {Bindings} bindings - The Bindings to consume
    */
-  public consumeBindings(bindings: Bindings): void {
+  public consumeBindings(bindings: Bindings): Promise<void> {
+    this.waitCounter++;
+
     // Select the bindings on which we group
     const grouper = bindings
       .filter((_, variable: string) => this.groupVariables.has(variable))
       .toMap();
     const groupHash = this.hashBindings(grouper);
-
     // First member of group -> create new group
-    let group: IGroup | undefined = this.groups.get(groupHash);
-    if (!group) {
-      // Initialize state for all aggregators for new group
-      const aggregators: Record<string, AggregateEvaluator> = {};
-      for (const aggregate of this.pattern.aggregates) {
-        const key = termToString(aggregate.variable);
-        aggregators[key] = new AggregateEvaluator(aggregate, this.sparqleeConfig);
-        aggregators[key].put(bindings);
-      }
+    const groupTest: Promise<IGroup> | undefined = this.groupsInitializer.get(groupHash);
 
-      group = { aggregators, bindings: grouper };
-      this.groups.set(groupHash, group);
+    let res: Promise<any>;
+    if (!groupTest) {
+      // Initialize state for all aggregators for new group
+      const groupPromise = (async() => {
+        const aggregators: Record<string, AsyncAggregateEvaluator> = {};
+        await Promise.all(this.pattern.aggregates.map(async aggregate => {
+          const key = termToString(aggregate.variable);
+          aggregators[key] = new AsyncAggregateEvaluator(aggregate, this.sparqleeConfig);
+          await aggregators[key].put(bindings);
+        }));
 
-      if (this.distinctHashes) {
-        const bindingsHash = this.hashBindings(bindings);
-        this.distinctHashes.set(groupHash, new Set([ bindingsHash ]));
-      }
+        if (this.distinctHashes) {
+          const bindingsHash = this.hashBindings(bindings);
+          this.distinctHashes.set(groupHash, new Set([ bindingsHash ]));
+        }
+        return { aggregators, bindings: grouper };
+      })();
+      this.groupsInitializer.set(groupHash, groupPromise);
+      res = groupPromise.then(group => {
+        this.groups.set(groupHash, group);
```

Comment:
Instead of doing this via a separate `.then` call, could we just move this into the actual initializer promise a few lines above?

Reply:
I suppose we could yes.